### PR TITLE
 Default to using program files if where.exe doesn't find dotnet.exe

### PR DESCRIPTION
**Pull request by jkotalik**
Per discussion with @shirhatti and @glennc, we will always fallback to ProgramFiles if where.exe doesn't find dotnet.exe. This is for the azure websites scenario where we are considering not putting the x64 version of dotnet on the path.
@muratg we want this for preview1 in case we don't get the chance to update Azure websites scenarios before preview2.

### CODE REVIEW

File: src/CommonLib/hostfxr_utility.cpp [hunk 1]
```diff
@@ -429,105 +429,130 @@ HOSTFXR_UTILITY::FindDotnetExePath(
...
-        if (FAILED(hr = struDotnetSubstring.Copy(struDotnetLocationsString.QueryStr(), index - prevIndex)))
+        if (FAILED(struDotnetSubstring.Resize(dwPathSize)))
         {
             goto Finished;
         }
-        prevIndex = index;
 
-        if (!GetBinaryTypeW(struDotnetSubstring.QueryStr(), &dwBinaryType))
+        // Program files will changes based on the 
+        dwNumBytesRead = GetEnvironmentVariable(L"ProgramFiles", struDotnetSubstring.QueryStr(), dwPathSize);
```
Comment (shirhatti) [inline src/CommonLib/hostfxr_utility.cpp]: You're only trying 64-bit fallback?

Comment (davidfowl) [reply src/CommonLib/hostfxr_utility.cpp]: Shouldn't it look in the appropriate place based on the bitness?

Comment (Tratcher) [reply src/CommonLib/hostfxr_utility.cpp]: %ProgramFiles% is bitness aware

Comment (davidfowl) [reply src/CommonLib/hostfxr_utility.cpp]: Ah I thought this was the path, I didn't see the GetEnvironmentVariable. Hopefully we have tests for this 😄

Comment (jkotalik) [reply src/CommonLib/hostfxr_utility.cpp]: Yeah as @Tratcher said, PF is bitness aware. 

@davidfowl native tests are coming soon ™️

Comment (jkotalik) [reply src/CommonLib/hostfxr_utility.cpp]: image
versus
image

File: src/CommonLib/hostfxr_utility.cpp [hunk 2]
```diff
@@ -429,105 +429,130 @@ HOSTFXR_UTILITY::FindDotnetExePath(
...
-    // where.exe will return 1 if the file is not found
+    // where.exe will return 0 on success, 1 if the file is not found
     // and 2 if there was an error. Check if the exit code is 1 and set
     // a new hr result saying it couldn't find dotnet.exe
     // 
-    if (!GetExitCodeProcess(processInformation.hProcess, &dwExitCode) ||
-        dwExitCode != 0)
+    if (!GetExitCodeProcess(processInformation.hProcess, &dwExitCode))
+    {
+        hr = HRESULT_FROM_WIN32(GetLastError());
```
Comment (pan-wang) [inline src/CommonLib/hostfxr_utility.cpp]: fallback instead of finish?

File: src/CommonLib/hostfxr_utility.cpp [hunk 3]
```diff
@@ -429,105 +429,130 @@ HOSTFXR_UTILITY::FindDotnetExePath(
...
-    }
-    else
-    {
-        SYSTEM_INFO systemInfo;
-        GetNativeSystemInfo(&systemInfo);
-        fIsCurrentProcess64Bit = systemInfo.wProcessorArchitecture == PROCESSOR_ARCHITECTURE_AMD64;
+            if (GetBinaryTypeW(struDotnetSubstring.QueryStr(), &dwBinaryType) &&
+                fIsCurrentProcess64Bit == (dwBinaryType == SCS_64BIT_BINARY)) {
+                // Found a valid dotnet.
+                struDotnetPath->Copy(struDotnetSubstring);
```
Comment (pan-wang) [inline src/CommonLib/hostfxr_utility.cpp]: copy may fail. please check hr

File: src/CommonLib/hostfxr_utility.cpp [hunk 4]
```diff
@@ -429,105 +429,130 @@ HOSTFXR_UTILITY::FindDotnetExePath(
...
+    // Look in ProgramFiles
     while (!fFound)
     {
-        index = struDotnetLocationsString.IndexOf(L"\r\n", prevIndex);
-        if (index == -1)
-        {
-            break;
-        }
-        if (FAILED(hr = struDotnetSubstring.Copy(struDotnetLocationsString.QueryStr(), index - prevIndex)))
+        if (FAILED(struDotnetSubstring.Resize(dwPathSize)))
```
Comment (pan-wang) [inline src/CommonLib/hostfxr_utility.cpp]: missing hresult

File: src/CommonLib/hostfxr_utility.cpp [hunk 5]
```diff
@@ -429,105 +429,130 @@ HOSTFXR_UTILITY::FindDotnetExePath(
...
+                FAILED(hr = struDotnetSubstring.Append(L"\\dotnet\\dotnet.exe")))
+            {
+                goto Finished;
+            }
+            if (!UTILITY::CheckIfFileExists(struDotnetSubstring.QueryStr()))
+            {
+                hr = HRESULT_FROM_WIN32( GetLastError() );
+                goto Finished;
+            }
             struDotnetPath->Copy(struDotnetSubstring);
```
Comment (pan-wang) [inline src/CommonLib/hostfxr_utility.cpp]: Copy may fail here

File: src/CommonLib/hostfxr_utility.cpp [hunk 6]
```diff
@@ -429,105 +429,130 @@ HOSTFXR_UTILITY::FindDotnetExePath(
...
             fFound = TRUE;
         }
     }
 
-    if (!fFound)
```
Comment (pan-wang) [inline src/CommonLib/hostfxr_utility.cpp]: Do we have window event log for not being able to find dotnet.exe

Comment (jkotalik) [reply src/CommonLib/hostfxr_utility.cpp]: I believe it is part of this PR pending a bit of refactoring: https://github.com/aspnet/IISIntegration/pull/533/files#diff-093466e281c521516f40989e2d5cb1caR302

**Comment by jkotalik** [comment]
I think we will wait on this until preview 2, correct @shirhatti ?

**Pull request closed without merging**